llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp

```diff
@@ -996,6 +997,150 @@ Instruction *InstCombinerImpl::foldPHIArgOpIntoPHI(PHINode &PN) {
   return NewCI;
 }
 
+/// Try to fold reduction ops interleaved through two PHIs to a single PHI.
+///
+/// For example, combine:
+///   %phi1 = phi [init1, %BB1], [%op1, %BB2]
+///   %phi2 = phi [init2, %BB1], [%op2, %BB2]
+///   %op1 = binop %phi1, constant1
+///   %op2 = binop %phi2, constant2
+///   %rdx = binop %op1, %op2
```

Comment:
Why does the pattern start matching at one of the phi nodes rather than at %rdx? This is pretty unusual for InstCombine, and it's not immediately obvious to me why it is necessary.

Reply:
> Would it be preferable to move the pattern to a method similar to foldBinopWithPhiOperands and start the match from %rdx? Or do you have something else in mind? :)

It would be better to handle this pattern in a separate helper like `foldBinopWithRecurrence`, then call it inside `foldBinopWithPhiOperands`. This approach may reuse some existing checks and avoid adding calls to `InstCombinerImpl::visit[BinOp]`.

Reply:
Hi @dtcxzyw, @nikic, I've moved the pattern to `foldBinopWithRecurrence` - please let me know if this is what you had in mind. Thanks!